hazelcast/src/test/java/com/hazelcast/util/FilteringClassLoader.java

```diff
@@ -85,49 +79,52 @@ public InputStream getResourceAsStream(String name) {
     protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
         isNotNull(name, "name");
 
-        for (String excludePackage : excludePackages) {
-            if (name.startsWith(excludePackage)) {
+        for (String excludedPackage : excludePackages) {
+            if (name.startsWith(excludedPackage)) {
                 throw new ClassNotFoundException(name + " - Package excluded explicitly!");
             }
         }
 
         if (enforcedSelfLoadingPackage != null && name.startsWith(enforcedSelfLoadingPackage)) {
-            Closeable classLoadingMutex = mutexProvider.getMutexForClass(name);
-            try {
-                //noinspection SynchronizationOnLocalVariableOrMethodParameter
-                synchronized (classLoadingMutex) {
-                    Class<?> clazz = findLoadedClass(name);
-                    if (clazz == null) {
-                        clazz = loadAndDefineClass(name);
-                    }
-                    return clazz;
+            // we don't call registerAsParallelCapable() on JDK7+, so we need to synchronize on this.
+            synchronized (this) {
+                Class<?> clazz = findLoadedClass(name);
+                if (clazz == null) {
+                    clazz = loadAndDefineClass(name);
                 }
-            } finally {
-                closeResource(classLoadingMutex);
+                return clazz;
             }
         }
         return delegatingClassLoader.loadClass(name);
     }
 
     private Class<?> loadAndDefineClass(String name) throws ClassNotFoundException {
         InputStream is = null;
-        ByteArrayOutputStream os = null;
+        ByteArrayOutputStream2 os = null;
         try {
-            is = getResourceAsStream(pattern.matcher(name).replaceAll("/").concat(".class"));
-            os = new ByteArrayOutputStream();
+            is = getResourceAsStream(name.replace('.', '/') + ".class");
+            os = new ByteArrayOutputStream2();
 
             int length;
             while ((length = is.read(buffer)) != -1) {
                 os.write(buffer, 0, length);
             }
 
-            byte[] data = os.toByteArray();
-            return defineClass(name, data, 0, data.length);
+            return defineClass(name, os.getInternalBuffer(), 0, os.size());
         } catch (Exception e) {
             throw new ClassNotFoundException(name, e);
         } finally {
             closeResource(os);
             closeResource(is);
         }
     }
+
+    /**
+     * {@link ByteArrayOutputStream} happy to share its internal buffer.
+     */
+    private static class ByteArrayOutputStream2 extends ByteArrayOutputStream {
```

Comment:
Can we call this `TestByteArrayOutputStream`? I think that would be better suited to our latest test cleanups.